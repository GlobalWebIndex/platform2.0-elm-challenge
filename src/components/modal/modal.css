#modal-full-screen {
  position: fixed;
  z-index: 1;
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: rgba(0, 0, 0, 0.7);
}

#modal-container {
  position: absolute;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: stretch;
}

#modal-image {
  height: 80vh;
}

#modal-footer {
  display: flex;
  justify-content: space-between;
  width: 100%;
  background-color: white;
}
