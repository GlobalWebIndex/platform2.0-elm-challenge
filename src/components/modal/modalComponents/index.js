import Footer from './footer/footer';
import CloseButton from './closeButton/closeButton';

export {
  Footer,
  CloseButton,
};
