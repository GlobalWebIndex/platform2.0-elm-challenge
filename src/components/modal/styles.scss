@import '../../assets/stylesheets/classes.scss';

.dialog {
  $actionsHeight: 50px;

  width: 680px;
  height: 90vh;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  gap: 2rem;

  &::backdrop {
    background-color: rgba(0, 0, 0, 0.6);
  }

  .container {
    padding: 1rem;
    overflow: auto;
    box-sizing: border-box;
    height: calc(100% - #{$actionsHeight});
  }

  .actions {
    height: $actionsHeight;
    border-top: 1px solid black;
    position: absolute;
    width: 100%;
    left: 0;
    bottom: 0;
    padding: 2rem;
    box-sizing: border-box;
    background-color: white;
    display: flex;
    justify-content: flex-end;
    align-items: center;
  }

  button {
    @extend %baseButton;

    background-color: transparent;
    border: 1px solid #6c757d;
    color: #6c757d;

    &:hover {
      color: #fff;
      background-color: #6c757d;
      border-color: #6c757d;
    }
  }
}

@media screen and (min-width: 576px) {
  .dialog {
    .actions {
      button {
        width: 20%;
      }
    }
  }
}
