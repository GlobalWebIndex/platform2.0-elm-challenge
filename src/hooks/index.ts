import { useEffect, useState } from 'react';
import { useLocation } from 'react-router-dom';

export function useData<T>(
  callback: () => Promise<T>,
  handleError?: () => void
) {
  const [data, setData] = useState<T | null>(null);

  useEffect(() => {
    async function fetchData() {
      try {
        const payload = await callback.call(null);

        setData(payload);
      } catch (error) {
        handleError?.();
      }
    }

    if (data === null) fetchData();
    // eslint-disable-next-line
  }, []);

  // Need this specific code structure
  // to achieve the desired typing
  if (data === null) {
    return [false, null] as const;
  }

  return [true, data] as const;
}

export function usePageTitle() {
  const location = useLocation();
  const title = location.pathname.substring(1);
  const defaultTitle = 'My Cat App';

  return title === '' || title.includes('images') ? defaultTitle : title;
}
