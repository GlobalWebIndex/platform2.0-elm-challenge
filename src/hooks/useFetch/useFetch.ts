import {useCallback, useEffect, useMemo, useState} from 'react';

export const useFetch = <T>(
  endpoint: string
): {abort: () => void; data: T | null; error: Error | null; isFetching: boolean} => {
  const [data, setData] = useState<T | null>(null);
  const [isFetching, setIsFetching] = useState(true);
  const [error, setError] = useState<Error | null>(null);

  const abortController = useMemo(() => new AbortController(), []);

  useEffect(() => {
    fetch(`https://api.thecatapi.com/v1/${endpoint}`, {
      signal: abortController.signal
    })
      .then(response => {
        if (!response.ok) {
          throw new Error('Failed to fetch data');
        }

        return response.json() as T;
      })
      .then(response => {
        setData(response);
      })
      .catch((e: unknown) => {
        if ((e as Error).name === 'AbortError') {
          return;
        }

        setError(e as Error);
      })
      .finally(() => {
        setIsFetching(false);
      });

    return (): void => {
      abortController.abort();
    };
  }, [abortController, endpoint]);

  const abort = useCallback<() => void>(() => {
    abortController.abort();
  }, [abortController]);

  return {abort, data, error, isFetching};
};
