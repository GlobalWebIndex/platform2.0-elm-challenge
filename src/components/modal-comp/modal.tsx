"use client";

import { SyntheticEvent, useEffect, useRef } from "react";
import styles from "./modal.module.scss";
import { MODAL_PROPS } from "@/lib/types";
import { usePathname, useSearchParams, useRouter } from "next/navigation";

export default function Modal({ title, children }: MODAL_PROPS) {
  const dialogRef = useRef<null | HTMLDialogElement>(null);
  const searchParams = useSearchParams();
  const pathname = usePathname();
  const router = useRouter();
  const imageId = searchParams.get("imageId");
  const breedId = searchParams.get("breedId");

  //We need to show the modal in two cases.
  //The first is when we need to show the detail of a specific breed.
  //The second is when we need to show cat images of a specific breed.
  const showModal =
    //case 1
    (imageId && pathname === "/") ||
    //case 2
    (breedId && pathname === "/breeds");

  useEffect(() => {
    if (showModal) {
      showDialog();
    } else {
      closeDialog();
    }
  }, [showModal]);

  function onCloseDialog(e: SyntheticEvent<HTMLDialogElement, Event>) {
    router.push(pathname);
  }

  function closeDialog() {
    dialogRef.current?.close();
  }

  function showDialog() {
    dialogRef.current?.showModal();
  }

  return (
    <dialog onClose={onCloseDialog} className={styles.dialog} ref={dialogRef}>
      <div className={styles.container}>
        <header>
          <h3 className="title">{title}</h3>
          <button onClick={closeDialog} className="closeBtn">
            X
          </button>
        </header>
        <main>{children}</main>
        <footer>
          <button onClick={closeDialog} className={`${styles.round} closeBtn`}>
            Close
          </button>
        </footer>
      </div>
    </dialog>
  );
}
