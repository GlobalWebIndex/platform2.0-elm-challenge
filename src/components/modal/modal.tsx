import React, {useEffect, useRef} from 'react';
import {useLocation, useNavigate, useParams} from 'react-router';
import type {BreedImages} from 'cat-lover/types/breeds';
import type {Cat} from 'cat-lover/types/cats';
import {useFetch} from '../../hooks/useFetch';
import {BreedDetails} from '../breedDetails';
import {CatDetails} from '../catDetails';
import {Loader} from '../loader';
import styles from './styles.scss';

export const Modal = (): JSX.Element => {
  const navigate = useNavigate();
  const {pathname} = useLocation();
  const dialogRef = useRef<HTMLDialogElement>(null);
  const {id} = useParams<'id'>();

  const isInBreedView = pathname.includes('breeds');

  const {data, isFetching} = useFetch<BreedImages | Cat>(
    isInBreedView ? `images/search?breed_ids=${id as string}` : `images/${id as string}`
  );

  useEffect(() => {
    if (dialogRef.current) {
      dialogRef.current.showModal();
    }
  }, []);

  const handleClose = (): void => {
    (dialogRef.current as HTMLDialogElement).close();

    navigate(-1);
  };

  return (
    <dialog className={styles.dialog} ref={dialogRef} role="dialog">
      {isFetching ? (
        <Loader />
      ) : (
        <>
          <div className={styles.container}>
            {isInBreedView ? <BreedDetails breedImages={data as BreedImages} /> : <CatDetails cat={data as Cat} />}
          </div>
          <div className={styles.actions}>
            <button onClick={handleClose}>Back</button>
          </div>
        </>
      )}
    </dialog>
  );
};
