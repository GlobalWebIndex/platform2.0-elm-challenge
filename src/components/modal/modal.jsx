import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import * as reduxActions from '../../redux/actions';
import { Footer, CloseButton } from './modalComponents';
import './modal.css';

const addEscKeyEventListener = (callback) => {
  const handler = (event) => {
    if (event.keyCode === 27) {
      callback();
    }
    document.removeEventListener('keydown', handler);
  };
  document.addEventListener('keydown', handler);
};

const Modal = () => {
  const modalProps = useSelector((state) => state.modalProps);
  const dispatch = useDispatch();

  addEscKeyEventListener(() => dispatch(reduxActions.updateModalPropsAction({})));

  return (
    <div
      id="modal-full-screen"
      data-testid="modal"
      style={modalProps.url ? { display: 'flex' } : { display: 'none' }}
    >
      <div id="modal-container">
        <CloseButton clickHandler={() => dispatch(reduxActions.updateModalPropsAction({}))} />
        <img
          id="modal-image"
          src={modalProps.url}
          alt="cat_modal"
        />
        <Footer
          modalProps={modalProps}
          toggleFavourite={() => {
            if (modalProps.favouriteId) {
              dispatch(reduxActions.deleteFavouriteAction(modalProps.favouriteId));
            } else {
              dispatch(reduxActions.saveFavouriteAction(modalProps.id, modalProps.url));
            }
          }}
        />
      </div>
    </div>
  );
};

export default Modal;
