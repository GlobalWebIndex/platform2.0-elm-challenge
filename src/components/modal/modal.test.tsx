import React from 'react';
import {MemoryRouter} from 'react-router-dom';
import {render, screen} from '@testing-library/react';
import {userEvent} from '@testing-library/user-event';
import type {BreedImages} from 'cat-lover/types/breeds';
import type {Cat} from 'cat-lover/types/cats';
import * as UseFetch from '../../hooks/useFetch';
import {Modal} from './modal';

type UseFetch = typeof UseFetch.useFetch<BreedImages | Cat>;
type UseFetchModule = Omit<typeof UseFetch, 'useFetch'> & {useFetch: UseFetch};

const mockUseFetch = jest.fn<ReturnType<UseFetch>, Parameters<UseFetch>>();
jest.mock<UseFetchModule>('../../hooks/useFetch', () => ({
  useFetch: (...args) => mockUseFetch(...args)
}));

describe('<Modal />', () => {
  const mockClose = jest.fn();
  const mockShowModal = jest.fn();

  beforeAll(() => {
    HTMLDialogElement.prototype.close = mockClose;
    HTMLDialogElement.prototype.showModal = mockShowModal;
  });

  beforeEach(() => {
    mockUseFetch.mockReturnValue({
      abort: jest.fn(),
      data: [],
      error: null,
      isFetching: false
    });
  });

  it('should open the dialog', () => {
    render(
      <MemoryRouter initialEntries={['/cats/1']}>
        <Modal />
      </MemoryRouter>
    );

    expect(mockShowModal).toHaveBeenCalledTimes(1);
  });

  describe('when data is fetching', () => {
    beforeEach(() => {
      mockUseFetch.mockReturnValue({
        abort: jest.fn(),
        data: null,
        error: null,
        isFetching: true
      });
    });

    it('should render Loader', () => {
      const {container} = render(
        <MemoryRouter initialEntries={['/cats/1']}>
          <Modal />
        </MemoryRouter>
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('when data is fetched for cat', () => {
    beforeEach(() => {
      const mockCat = {
        breeds: [],
        height: 500,
        id: '',
        url: 'http://example.com/cat.jpg',
        width: 400
      };

      mockUseFetch.mockReturnValue({
        abort: jest.fn(),
        data: mockCat,
        error: null,
        isFetching: false
      });
    });

    it('should render cat details', () => {
      const {container} = render(
        <MemoryRouter initialEntries={['/cats/1']}>
          <Modal />
        </MemoryRouter>
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('when data is fetched for breed', () => {
    beforeEach(() => {
      const mockBreedImages: BreedImages = [
        {
          height: 500,
          id: '1',
          url: 'http://example.com/cat.jpg',
          width: 400
        }
      ];

      mockUseFetch.mockReturnValue({
        abort: jest.fn(),
        data: mockBreedImages,
        error: null,
        isFetching: false
      });
    });

    it('should render breed details', () => {
      const {container} = render(
        <MemoryRouter initialEntries={['/breeds/1']}>
          <Modal />
        </MemoryRouter>
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('when the user clicks on the "Back" button', () => {
    beforeEach(() => {
      mockUseFetch.mockReturnValue({
        abort: jest.fn(),
        data: [],
        error: null,
        isFetching: false
      });
    });

    it('should close the modal', async () => {
      render(
        <MemoryRouter initialEntries={['/cats/1']}>
          <Modal />
        </MemoryRouter>
      );

      await userEvent.click(screen.getByText('Back'));

      expect(mockClose).toHaveBeenCalledTimes(1);
    });
  });
});
