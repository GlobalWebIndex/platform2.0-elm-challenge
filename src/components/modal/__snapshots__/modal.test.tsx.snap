// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> when data is fetched for breed should render breed details 1`] = `
<div>
  <dialog
    class="dialog"
    role="dialog"
  >
    <div
      class="container"
    >
      <div
        class="images"
      >
        <a
          href="/cats/1"
        >
          <img
            alt="Breed"
            src="http://example.com/cat.jpg"
          />
        </a>
      </div>
    </div>
    <div
      class="actions"
    >
      <button>
        Back
      </button>
    </div>
  </dialog>
</div>
`;

exports[`<Modal /> when data is fetched for cat should render cat details 1`] = `
<div>
  <dialog
    class="dialog"
    role="dialog"
  >
    <div
      class="container"
    >
      <div
        class="container"
      >
        <div
          class="horizontal withNoBreeds"
        >
          <div
            class="image"
          >
            <img
              alt="Cat"
              src="http://example.com/cat.jpg"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="actions"
    >
      <button>
        Back
      </button>
    </div>
  </dialog>
</div>
`;

exports[`<Modal /> when data is fetching should render Loader 1`] = `
<div>
  <dialog
    class="dialog"
    role="dialog"
  >
    <div
      class="loader"
    >
      <svg>
        <use
          href="function SvgIcons(props) {
  return React.createElement(
    'svg', 
    Object.assign({}, props, {'data-file-name': SvgIcons.name})
  );
}#icon-cw"
        />
      </svg>
    </div>
  </dialog>
</div>
`;
